### DISCUSSION TITLE
How to preload JS code and have it accessible in different isolates or contexts

### DISCUSSION THREAD

**Discussion by tuzzmaniandevil** [caoccao/Javet#169] [Q&A]
This may be an odd situation, But I'll give some background.

I work at multi-tenet SaaS company that allows people to create custom apps, etc using JavaScript. We currently use Nashorn. To save load time and memory we create a single Nashorn engine per app and that single engine is accessed by all accounts and we use a separate context per account to prevent one account from storing a global variable that can be accessed by another account.

With Nashorn on the way out we are looking at replacing it.

So what I need assistance with is a way to efficiently share code across multiple contexts or isolates in a V8 engine without unnecessarily duplicating memory usage.

From my research this is possible to have a V8 Isolate then creating different contexts but setting each context security key to a unique value using `v8::Context::SetSecurityToken()`

I have looked at GraalJS as well, But it has some performance issues and other limitations. So I'm hoping Javet is the way forward.

**Comment by caoccao** [comment]
Javet is a wonderful scripting engine for SaaS supporting multi-tenancy with complete data segregation among the tenants. Here are my 2 cents for your reference.

> Nashorn

The memory or threading model in Nashorn is off the ECMAScript spec so that there are many undefined behaviors under race condition. I guess (please correct me if I'm wrong) in your case your application only allows read access to the shared data to avoid those gray area. Both Javet and GraalJS don't leave room for those undefined behaviors. That implies you need to find a new way, of course, this discussion as a beginning.

> GraalJS

I my humble opinion, it concerns me on the following.

- GraalJS is quite slow. The nature of SaaS applications is the custom scripts are mostly likely a one-time deal so that one million rounds of warm-up for best performance is ridiculous. V8 performs much better in the first round.
- GraalJS is slightly off the ECMAScript spec. I suspect there are thousands of minor differences between GraalJS and V8. Some critical feature mismatches would cost millions of loss, I think.
- GraalJS community edition is not at enterprise level. Of course, Oracle wants you to use GraalJS enterprise edition which could be very expensive.

> Multiple V8 Contexts

The execution among multiple V8 contexts in one V8 isolate is single-threaded. I don't think that's what you expect, please correct me if I'm wrong.

# Suggestion

My suggestion is Javet Engine Pool + ES Module.

## Javet Engine Pool

Javet Engine Pool works like a DB connection pool. It allows multiple tenants to reuse idle V8 isolates without creating/destroying V8 isolates frequently.

## ES Modules

ESM can be used to cache shared business logic and data among the tenants. Those ES modules are actively cached by V8.

# Pros and Cons

## Pros

- Best performance
- Complete data segregation
- Simple design

## Cons

- Additional memory footprint

Hope these make sense. You may visit [discord](https://discord.gg/R4vvKU96gw) to discuss with me in real-time if you have any further questions.
